src/net/http/UploadFile.cpp

```diff
@@ -0,0 +1,255 @@
+  AtScopeExit(&request) { request.StopIndirect(); };
+
+  request.SetFailOnError();
+
+  if (username != nullptr)
+    request.SetOption(CURLOPT_USERNAME, username);
+  if (password != nullptr)
+    request.SetOption(CURLOPT_PASSWORD, password);
+
+  env.SetCancelHandler([&]{
+    request.StopIndirect();
+    handler.Cancel();
+  });
+
+  AtScopeExit(&env) { env.SetCancelHandler({}); };
+
+  request.StartIndirect();
+  handler.Wait();
+
+  if (env.IsCancelled())
+    return -1;
+
+  return handler.GetReceived();
+}
+
+size_t
+Net::UploadFile(CurlGlobal &curl, const char *url, const TCHAR *username,
+  const TCHAR *password, const Path path, char *msg_buffer, size_t msg_buffer_size,
+  http_post *formpost, OperationEnvironment &env)
+{
+  assert(url != nullptr);
+  assert(path != nullptr);
+
+  return ::UploadFile(curl, url, username, password, msg_buffer, msg_buffer_size, 
+   formpost, env);
+
+}
+/**
+ * Writes one line of NMEA data.
+ *
+ * @param port the port to write to
+ * @param line the line without asterisk, checksum and newline
+ * @return true on success
+ */
+
+/**
+* @param CurlGlobal _curl - global curl component
+* @param const char *_url - server url for uploading file (with post)
+* @param const Path _path - filepath to the file to uploading
+* @param char *_msg_buffer    - message buffer for response data from server
+* @param size_t _msg_buffer_size - max size of response data  in bytes
+*/
+
+Net::UploadFileJob::UploadFileJob(CurlGlobal& _curl, const char *_url,
+  const Path _filepath, char *_msg_buffer, const size_t _msg_buffer_size)
+  : curl(_curl), url(_url), filepath(_filepath), msg_buffer(_msg_buffer),
+  msg_buffer_size(_msg_buffer_size)
+{
+  char curl_filename[0x200];
+  std::snprintf(curl_filename, sizeof(curl_filename), "%s",
+    (const char*)NarrowPathName(filepath));
+
+  LogFormat("UploadFile: %s", curl_filename);
+  curl_formfree(formpost);
+  curl_formadd(&formpost,
+    &lastptr,
+    CURLFORM_COPYNAME, "file",
+    CURLFORM_FILE, curl_filename,
+    CURLFORM_CONTENTTYPE, "application/octet-stream",
+    CURLFORM_END);
+}
+
+void
+Net::UploadFileJob::Run(OperationEnvironment &env)
+{
+  length = UploadFile(curl, url, username, password, filepath, msg_buffer, msg_buffer_size, formpost, env);
+}
+
+void
+Net::UploadFileJob::AddPostValue(const char *name, const char *value) {
+  curl_formadd(&formpost,
+    &lastptr,
+    CURLFORM_COPYNAME, name,
+    CURLFORM_COPYCONTENTS, value,
+    CURLFORM_END);
+}
+
+#ifdef _UNICODE
+void
+Net::UploadFileJob::AddPostValue(const char *name, const TCHAR *_value) {
+  NarrowString<0x200> value;
+  value.SetASCII(_value);
+  AddPostValue(name, value);
+}
+#endif  // _UNICODE
+
+void
+Net::UploadFileJob::AddPostValue(const char *name, const uint32_t value) {
+  AddPostValue(name, std::to_string(value).c_str());
+}
+
+void
+Net::UploadFileJob::AddPostValue(const char *name, const BrokenDate& value) {
+    char buffer[16];
+    FormatISO8601(buffer, value);
```

Comment:
That is an application-level protocol assumption. Such an assumption does not belong in this library.

Reply:
Ok, this I have removed, BrokenDate is unknown in this low-level component!